server/chat-plugins/chatlog.ts

```diff
@@ -109,12 +110,25 @@ const LogReader = new class {
 		return {official, normal, hidden, secret, deleted, personal, deletedPersonal};
 	}
 
-	async read(roomid: RoomID, day: string) {
-		const month = day.slice(0, -3);
-		const log = FS(`logs/chat/${roomid}/${month}/${day}.txt`);
-		if (!await log.exists()) return null;
-		const text = await log.read();
-		return text;
+	async read(roomid: RoomID, day: string, limit: number) {
+		const roomLog = await LogReader.get(roomid);
+		const stream = await roomLog!.getLog(day);
+		let buf = '';
+		let i = LogViewer.results || 0;
+		if (!stream) {
+			buf += `<p class="message-error">Room "${roomid}" doesn't have logs for ${day}</p>`;
+		} else {
+			let line;
+			while ((line = await stream.readLine()) !== null) {
+				const rendered = LogViewer.renderLine(line);
+				if (i++ > limit) return buf;
```

Comment:
You need to use `++i` here or else it increments after comparing. Also why are you incrementing here and on line 127?

A wiser order might be render -> add to buf if rendered & increment i ->if i >= limit break -> next loop